### PR TITLE
[POC] Improve MinIO support

### PR DESCRIPTION
**Pull request by chibenwa**
Storing all blobs into a single bucket hurts: impossible to run the GC or do S3 administrative operations

(minio underlying model is mapping to OS folders, so Tmail storing all in one bucket litterally creates folders with dozens of millions of files which is a VERY bad idea)

First commit adds a little handy abstraction allowing alternative implems to GenerationAwareBlobId and makes the system more open: NICE.

The second commit could be done in Twake mail if needed though I do know some James users relying on MinIO in the wild out there...

### CODE REVIEW

File: server/blob/blob-storage-strategy/src/main/java/org/apache/james/server/blob/deduplication/MinIOGenerationAwareBlobId.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        @Inject
+        public Factory(Clock clock, GenerationAwareBlobId.Configuration configuration, BlobId.Factory delegate) {
+            this.clock = clock;
+            this.configuration = configuration;
+            this.delegate = delegate;
+        }
+
+        @Override
+        public BlobId parse(String id) {
+            int separatorIndex1 = id.indexOf('/');
```
Comment (vttranlina) [inline server/blob/blob-storage-strategy/src/main/java/org/apache/james/server/blob/deduplication/MinIOGenerationAwareBlobId.java:55]: An alternative approach is to use the "family" character to identify the blobId: 
- family = 0 ->  delegate 
- family = 1 -> GenerationAwareBlobId 
- family = 2 -> MinIOGenerationAwareBlobId.

File: server/blob/blob-storage-strategy/src/main/java/org/apache/james/server/blob/deduplication/MinIOGenerationAwareBlobId.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            if (separatorIndex1 == -1 || separatorIndex1 == id.length() - 1) {
+                return decorateWithoutGeneration(id);
```
Comment (vttranlina) [inline server/blob/blob-storage-strategy/src/main/java/org/apache/james/server/blob/deduplication/MinIOGenerationAwareBlobId.java:57]: I understand this line aims to support compatibility, 
but it should correctly be `GenerationAwareBlobId.Factory.parse(id)`

File: server/blob/blob-storage-strategy/src/main/java/org/apache/james/server/blob/deduplication/MinIOGenerationAwareBlobId.java (hunk 1 above)
Comment (chibenwa) [reply server/blob/blob-storage-strategy/src/main/java/org/apache/james/server/blob/deduplication/MinIOGenerationAwareBlobId.java:55]: TBH less flexible: what if I already incremented several time the family?
Which value to use as a discriminant?

File: server/blob/blob-storage-strategy/src/main/java/org/apache/james/server/blob/deduplication/MinIOGenerationAwareBlobId.java (hunk 2 above)
Comment (chibenwa) [reply server/blob/blob-storage-strategy/src/main/java/org/apache/james/server/blob/deduplication/MinIOGenerationAwareBlobId.java:57]: No not really.

With MinIO running GC on the screwed up generations held with all files in a single bucket is anyway a dead end. I prefer using plain blobIds for it.

**Comment by Arsnael** [comment]
Just added the fixup for the quota-ratio, cause that's what the CI build is failing about

File: server/container/guice/blob/deduplication-gc/src/main/java/org/apache/james/modules/blobstore/BlobDeduplicationGCModule.java [hunk 3]
```diff
@@ -68,7 +70,14 @@ protected void configure() {
     @Singleton
     @Provides
     public BlobId.Factory generationAwareBlobIdFactory(Clock clock, PlainBlobId.Factory delegate, GenerationAwareBlobId.Configuration configuration) {
-        return new GenerationAwareBlobId.Factory(clock, delegate, configuration);
+            String property = System.getProperty("james.s3.minio.compatibility.mode");
```
Comment (vttranlina) [inline server/container/guice/blob/deduplication-gc/src/main/java/org/apache/james/modules/blobstore/BlobDeduplicationGCModule.java:73]: I see in the ticket 
```
Have an option in blobstore.properties to actually switch to that MinIO compatibility mode:
```
Do you need it? I can adapt it

**Pull request closed without merging**